### ISSUE TITLE
[question] Delay dependency graph resolution until after build-time?

### ISSUE THREAD

**Issue by mildsenseofdoom** [conan-io/conan#14160]
## Problem

We're building several conan packages for distribution, some of which we want to depend on other ones that we're building. I've tried adding these as `self.requires()` in `requirements()` as per this link: https://docs.conan.io/2/tutorial/creating_packages/add_dependencies_to_packages.html

However, when it comes time to build the packages, conan seems to require that it can find the dependencies when the package is built (i.e. I can't seem to simply promise a package will be available in the future) - this works for packages we happen to build first, but if we want to guarantee this then we'd have to essentially build a dependency graph at build-time to make sure that packages are built in the correct order, which is a lot of work for us.

## Possible Solution?

I'm hoping there's some way to tell conan to delay resolving the dependencies, and just resolve them at install-time? We're testing that we can install packages after we build them, so I'm confident that any missing dependencies would get spotted.

Another way might be to be able to add dependencies after the fact - i.e. build the package with no dependencies, and then add them - I can't think how I'd start going about that though.

Any help appreciated!

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @johngavingraham 

It is not completely clear to me, what is your intended flow, or the problem you are trying to solve. Maybe the "resolve dependencies" is a bit generic, so let me outline the process of installing dependencies:

When you do ``conan install`` this happens:

- First, the dependency graph of only "recipes" is computed. Which recipes, which versions, resolve version ranges, resolve latest revisions, etc. This requires evaluation of settings and options to resolve conditional ``requires`` in the graph
- Second step, once the full graph is computed, is calculating the ``package_id`` of every node in the graph, and then checking if they exist, they are installed in the cache, they are missing or if they would be built from source depending on ``--build=xxx`` argument.
- Finally, the resulting graph, with the computed ``package_id`` and origin (cache, build, download) is "installed" in the cache, packages are downloaded, built, etc.

Maybe the ``conan graph info`` command is what you are looking for. It only executes the first 2 steps, but it doesn't execute the final installation of the package binaries, then it doesn't need to download the binaries at all. There are other useful subcommands there, like ``conan graph build-order`` that returns an ordered list-of-lists of the things that need to be build in the right order of execution, for parallel distribution in a distributed CI, for example.

However there are some confusing things in your comment, like it is not possible to build a package without the dependencies first, how that would work? If a package ``mypkg`` has some ``#include "otherpkg/header.h"``, unless the ``otherpkg/someversion`` is actually installed by Conan and made available to the build, the the build will obviously fail, because the headers of ``otherpkg`` will not be available.

**Comment by mildsenseofdoom** [comment]
Thanks for the response @memsharded !

To clear up the confusion:

> When you do `conan install` this happens:

This may be the source of confusion! To be clear, this is when I'm doing a `conan build` - i.e. I get something like this (mildly sensitive information replaced, this is if I'm trying to build `CURRENT-PACKAGE` and that requires `REQUIRED-PACKAGE`):

```
$ conan build /tmp/tmp.MtiwYn5ZoJ

[[snip...]]

======== Computing dependency graph ========
REQUIRED-PACKAGE/VERSION: Not found in local cache, looking in remotes...
REQUIRED-PACKAGE/VERSION: Checking remote: conancenter
Graph root
    conanfile.py (CURRENT-PACKAGE/VERSION): /tmp/tmp.MtiwYn5ZoJ/conanfile.py
ERROR: Package 'REQUIRED-PACKAGE/VERSION' not resolved: Unable to find 'REQUIRED-PACKAGE/VERSION' in remotes.
```

In the above case, I will be building `REQUIRED-PACKAGE` later, but it'd be very convenient to not have to work out that I need to do them in any specific order. (This is simple for two packages but we're automating the build of many packages.)

> However there are some confusing things in your comment, like it is not possible to build a package without the dependencies first, how that would work?

So for us (right now at least, this is a very initial stage of a project) we're not actually "building" anything - we're just packaging up some files (Google Protobuf files) that can be consumed later. We'll be adding pre-built binaries at a later stage, but even then we'll still want to distribute just the raw Protobuf files, as it's possible that some of our consumers will want to compile them/use them in ways we haven't anticipated - and even when we're distributing binaries, these will have been built by other parts of our build system by the time it gets to Conan-packaging-time (i.e. we'd just use Conan to distribute the binaries, not build them).

_N.B. I'm aware the above sounds a bit crazy, alternatives and such are welcome but fundamentally with our business requirements, this is where we've ended up for now, at least!_

In other words, I'm trying to `conan build` a bunch of packages in a batch, and some of them depend on each other, so I either need to (I think) work out their dependencies myself so I make sure I build them in the right order (which I'd rather avoid) or tell conan "this package requires XXX, but don't go looking for it now - just look for it at install-time".

**Comment by memsharded** [comment]
I am afraid it is still not very clear:

> This may be the source of confusion! To be clear, this is when I'm doing a conan build - i.e. I get something like this (mildly sensitive information replaced,

The ``conan build`` command is doing in practice a ``conan install`` followed by the execution of the ``build()`` method, that will call the build system, configure, make, etc, to build from source.
If the dependencies like ``REQUIRED-PACKAGE/VERSION`` are not installed, then, by definition it will be impossible that the curren ``build()`` will succeed, because it will not be able to build from source if ``REQUIRED-PACKAGE/VERSION`` is not there first. And if ``REQUIRED-PACKAGE/VERSION`` is not necessary for this package to build, then it is not a dependency! it shouldn't be a ``requires`` in the first place? 


> So for us (right now at least, this is a very initial stage of a project) we're not actually "building" anything - we're just packaging up some files (Google Protobuf files) that can be consumed later. We'll be adding pre-built binaries at a later stage, but even then we'll still want to distribute just the raw Protobuf files, as it's possible that some of our consumers will want to compile them/use them in ways we haven't anticipated - and even when we're distributing binaries, these will have been built by other parts of our build system by the time it gets to Conan-packaging-time (i.e. we'd just use Conan to distribute the binaries, not build them).

Well, Conan is quite flexible, I guess that you can easily do things like:

```python
options = {"nodeps": [True, False]}
default_options = {"nodeps": False}

def requirements(self):
      if not self.options.nodeps:
           self.requires("....)

def package_id(self):
     # not sure what to do here what would be the binary model???
```

> N.B. I'm aware the above sounds a bit crazy, alternatives and such are welcome but fundamentally with our business requirements, this is where we've ended up for now, at least!

Yes, a little bit 😅 

I am afraid that the use case is still not clear. Even if you are only packaging some files (it is possible, and we have seen users using Conan for packaging all kinds of artifacts, images, files, sources, kernels, etc), those files are still necessary to build the consumers, and if they are not, is because they are not a real ``requires`` of the package? 
I don't know if we will be able to understand the use case via Github comments, if you want to drop an email to info@conan.io we might connect via zoom to learn about your use case.

**Comment by mildsenseofdoom** [comment]
> The conan build command is doing in practice a conan install...

Ah, I did not realise this! That makes your previous response clearer for me, thanks :)

As for the problem, having a quick play, it seems that being able to have a `nodeps` option as you demonstrated is just the ticket - I can build the packages while ignoring dependencies, and when it comes to install time it does notice if I've specified a package that does not exist.

> I am afraid that the use case is still not clear. Even if you are only packaging some files... those files are still necessary to build the consumers, and if they are not, is because they are not a real requires of the package?

Yes, I don't think I've explained it very well! Files from different packages are necessary to build some consumers (some packages import files from other packages), it's not a problem of what packages are in what files that I was encountering, but purely a problem of build order - if A depended on B then I'd have to make sure that I build B first. This isn't an intractable problem, just one that it seemed I could get around in a simpler way. Or put another way, as we're just packaging artifacts then we don't need to access them to package them up and distribute them, but when our consumers get them we want Conan to automatically get what it needs.

Thanks very much for your help @memsharded :)

**Comment by memsharded** [comment]
> As for the problem, having a quick play, it seems that being able to have a nodeps option as you demonstrated is just the ticket - I can build the packages while ignoring dependencies, and when it comes to install time it does notice if I've specified a package that does not exist.

Ok, seems it is a bit hacky, but if it works for you, I am happy 🙂 


> Yes, I don't think I've explained it very well! Files from different packages are necessary to build some consumers (some packages import files from other packages), it's not a problem of what packages are in what files that I was encountering, but purely a problem of build order - if A depended on B then I'd have to make sure that I build B first. This isn't an intractable problem, just one that it seemed I could get around in a simpler way. Or put another way, as we're just packaging artifacts then we don't need to access them to package them up and distribute them, but when our consumers get them we want Conan to automatically get what it needs.

Oh, this is interesting information. It could be then (probably a bit hacky too, but interesting) approach in Conan 2.0 with the new traits. Let's say something like this:

```python
def requirements(self):
      active = False if self.options.nodeps else None  # None is the default
      self.requires("other/version", headers=active, libs=active, run=active)
      self.requires("other2/version", headers=active, libs=active, run=active)
```
This is a very rough idea that I haven't tested, but maybe worth investigating. This approach actually requires that the transitive dependencies to exist, as a recipe/version, but the binaries will be **Skip**, that is, they are not really needed, and Conan will not require downloading the binaries or building the binaries, they don't need to exist yet. Still, it will compute and report the dependency graph, which might be very useful information, recall that there is "binary model" that defines the ``package_id`` based on many factors, including the dependencies and the versions of the dependencies. This model might become valuable or even very necessary when users are consuming this graph to be able to consume the right binaries, built against the correct set of transitive dependencies.


Just an idea, I am a bit concerned that you would be digging yourself into a hole of complexity that might be painful in the future. In any case, if the above hints were enough for your use case, I am happy to have helped! :)
